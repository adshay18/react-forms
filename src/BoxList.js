import { useState } from 'react';
import Box from './Box';
import { v4 as uuid } from 'uuid';
import NewBoxForm from './NewBoxForm';

const BoxList = () => {
	const [ boxes, setBoxes ] = useState([]);

	const addBox = (box) => {
		let newBox = { ...box, id: uuid() };
		setBoxes((boxes) => [ ...boxes, newBox ]);
	};

	return (
		<div className="BoxList">
			<NewBoxForm addBox={addBox} />
			<h2>Boxes:</h2>
			{boxes.map((box) => (
				<Box
					key={box.id}
					backgroundColor={box.backgroundColor}
					width={Number(box.width)}
					height={Number(box.height)}
				/>
			))}
		</div>
	);
};

export default BoxList;
