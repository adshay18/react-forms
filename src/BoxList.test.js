import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import BoxList from './BoxList';

// Smoke test
it('renders without crashing', function() {
	render(<BoxList />);
});

// Snapshot
it('matches snapshot', function() {
	const { asFragment } = render(<BoxList />);
	expect(asFragment()).toMatchSnapshot();
});

it('can add a new box and delete it', function() {
	const { getByLabelText, queryByText } = render(<BoxList />);

	// no items yet
	expect(document.querySelector('div.Box')).not.toBeInTheDocument();

	const colorInput = getByLabelText('Color:');
	const widthInput = getByLabelText('Width:');
	const heightInput = getByLabelText('Height:');
	const submitBtn = queryByText('Add box');

	// fill out the form
	fireEvent.change(colorInput, { target: { value: '#000000' } });
	fireEvent.change(widthInput, { target: { value: 100 } });
	fireEvent.change(heightInput, { target: { value: 100 } });
	fireEvent.click(submitBtn);

	// item exists!
	expect(document.querySelector('div.Box')).toBeInTheDocument();

	const deleteBtn = queryByText('X');

	// click delete button
	fireEvent.click(deleteBtn);

	// box is deleted
	expect(document.querySelector('div.Box')).not.toBeInTheDocument();
});
