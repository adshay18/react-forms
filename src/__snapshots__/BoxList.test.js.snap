// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <form>
      <label
        for="backgroundColor"
      >
        Color: 
      </label>
      <input
        id="backgroundColor"
        name="backgroundColor"
        type="color"
        value="#000000"
      />
      <label
        for="width"
      >
        Width: 
      </label>
      <input
        id="width"
        name="width"
        type="number"
        value="0"
      />
      <label
        for="height"
      >
        Height: 
      </label>
      <input
        id="height"
        name="height"
        type="number"
        value="0"
      />
      <button>
        Add box
      </button>
    </form>
    <h2>
      Boxes:
    </h2>
  </div>
</DocumentFragment>
`;
